### PR TITLE
Add backspace unindents feature

### LINKED ISSUE #15180: [Feature request] Have backspace delete spaces back to previous tabstop
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description of the Issue

In a nutshell, implement [SCI_SETBACKSPACEUNINDENTS](https://www.scintilla.org/ScintillaDoc.html#SCI_SETBACKSPACEUNINDENTS)(true/false), with a new preference setting controlling the true/false part as an enable/disable.

Preference setting could go on the `Editing2` subpage, with checkbox text of `Enable Backspacing Unindent` or `Enable backspacing of multiple spaces to previous tabstop` or ...?

Enable by default (so that users realize that it is a new feature).

### Describe the solution you'd like.

See Description above.

### Debug Information

```shell
Notepad++ v8.6.7   (64-bit)
Build time : May 12 2024 - 20:04:00
Path : C:\NotepadPlusPlus\Misc\_NoBackup\Releases\npp.8.6.7\npp.8.6.7.portable.x64\notepad++.exe
Command Line : -titleAdd="8.6.7 64-bit" -multiInst
Admin mode : OFF
Local Conf mode : ON
Cloud Config : OFF
Periodic Backup : ON
OS Name : Windows 11 Pro (64-bit)
OS Version : 23H2
OS Build : 22631.3593
Current ANSI codepage : 1252
Plugins : 
    mimeTools (3.1)
    NppConverter (4.6)
    NppExport (0.4)
    PythonScript (3.0.16)
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by alankilborn**
Fix #15180 

Too much code changed to make me feel comfortable.
I got "talked into" making this feature language-specific.

### CODE REVIEW

File: PowerEditor/src/Parameters.h [hunk 1]
```diff
@@ -1034,6 +1035,7 @@ struct Lang final
 
 	bool _isTabReplacedBySpace = false;
 	int _tabSize = -1;
+	bool _isBackspaceUnindents = false;
```
Comment (donho) [inline PowerEditor/src/Parameters.h:1038]: Shouldn't it be `_isBackspaceUnindent` ?

File: PowerEditor/src/Parameters.h [hunk 2]
```diff
@@ -786,6 +786,7 @@ struct NppGUI final
 
 	int _tabSize = 4;
 	bool _tabReplacedBySpace = false;
+	bool _backspaceUnindents = false;
```
Comment (donho) [inline PowerEditor/src/Parameters.h:789]: Rather `_backspaceUnindent`.

File: PowerEditor/src/Parameters.cpp [hunk 3]
```diff
@@ -5307,6 +5308,10 @@ void NppParameters::feedGUIParameters(TiXmlNode *node)
 			val = element->Attribute(TEXT("replaceBySpace"));
 			if (val)
 				_nppGUI._tabReplacedBySpace = (!lstrcmp(val, TEXT("yes")));
+
+			val = element->Attribute(TEXT("backspaceUnindents"));
```
Comment (donho) [inline PowerEditor/src/Parameters.cpp:5312]: "backspaceUnindent"

File: PowerEditor/src/NppBigSwitch.cpp [hunk 4]
```diff
@@ -3626,6 +3626,7 @@ LRESULT Notepad_plus::process(HWND hwnd, UINT message, WPARAM wParam, LPARAM lPa
 
 		case NPPM_INTERNAL_SETTING_TAB_REPLCESPACE:
 		case NPPM_INTERNAL_SETTING_TAB_SIZE:
+		case NPPM_INTERNAL_SETTING_BACKSPACE_UNINDENTS:
```
Comment (donho) [inline PowerEditor/src/NppBigSwitch.cpp:3629]: Please change the following:
1. remove the definition of `NPPM_INTERNAL_SETTING_BACKSPACE_UNINDENTS` and comment the drfinition of `NPPM_INTERNAL_SETTING_TAB_SIZE`.
2. rename `NPPM_INTERNAL_SETTING_TAB_REPLCESPACE` to `NPPM_INTERNAL_SET_TAB_SETTINGS`.

File: PowerEditor/installer/nativeLang/english.xml [hunk 5]
```diff
@@ -1107,6 +1107,7 @@ Translation note:
 					<Item id="6311" name="Tab character"/>
 					<Item id="6510" name="Use default value"/>
 					<Item id="6335" name="Treat backslash as escape character for SQL"/>
+					<Item id="6512" name="Backspace key unindents instead of removing single space"/>
```
Comment (donho) [inline PowerEditor/installer/nativeLang/english.xml:1110]: Please don't forget custom_english.xml

**Review by donho — CHANGES_REQUESTED** [review]
Please consider the suggestions.

File: PowerEditor/src/Parameters.h [hunk 6]
```diff
@@ -1063,13 +1065,15 @@ struct Lang final
 		_pCommentEnd = commentEnd;
 	}
 
-	void setTabInfo(int tabInfo)
+	void setTabInfo(int tabInfo, bool bsInfo)
```
[the remark this answers is not in the document]
Comment (alankilborn) [reply PowerEditor/src/Parameters.h]: OK, changed all usages of `...unindents` to be `...unindent`.

File: PowerEditor/installer/nativeLang/english.xml (hunk 5 above)
Comment (alankilborn) [reply PowerEditor/installer/nativeLang/english.xml:1108]: I've never updated the "custom" version of the file before...but OK, I will do so this time and also moving forward.

**Comment by donho** [comment]
Sorry @alankilborn 
I have to close this PR due to the conflicts (you should based from the updated master locally then rebase and force push, but NEVER merge the master).
Please create a new branch and a new PR.

**Comment by alankilborn** [comment]
> you should based from the updated master locally then rebase and force push, but NEVER merge the master

Sorry...  I got an error message I had never seen before, and it appears I took what turned out to be "bad advice" while researching the solution (based on searching for that error msg).

It's easy enough for me to recreate the change in a new branch/PR.
I will do so soon.

**Comment by donho** [comment]
Thank you @alankilborn !

**Pull request closed without merging**